programming-language-for-tester-assignment/NganLe/intermediate-problems-1-to-27.js

```diff
@@ -0,0 +1,429 @@
+/**
+ * ## Intermediate Problems
+    1.  **Switch Statement**
+    - Use a switch statement to log a greeting based on the current day of the week.
+    - Explanation: Teaches the switch case control structure.
+ */
+let today = new Date().getDay();
+let greeting;
+switch (today) {
+    case 0:
+        greeting = 'Hello Sunday';
+        break;
+    case 1:
+        greeting = 'Hello Monday';
+        break;
+    case 2:
+        greeting = 'Hello Tuesday';
+        break;
+    case 3:
+        greeting = 'Hello Wednesday';
+        break;
+    case 4:
+        greeting = 'Hello Thursday';
+        break;
+    case 5:
+        greeting = 'Hello Friday';
+        break;
+    case 6:
+        greeting = 'Hello Saturday';
+        break;
+    default:
+        greeting = 'Invalid day';
+}
+console.log(greeting);
+
+/**
+ * 2.  **Ternary Operator**
+    - Write a one-liner that uses the ternary operator to log "Yes" if a variable is true, and "No" if false.
+    - Explanation: Introduces the ternary operator for conditional expressions.
+ */
+function checkVar(variable) {
+    return variable ? 'Yes' : 'No';
+}
+console.log(checkVar(true));
+
+/**
+ * 3. **Function Returning a Function**
+    - Write a function that returns another function which logs "Hello, World!".
+    - Explanation: Goes into higher-order functions and closures.
+ */
+function returnFunc() {
+    return function () {
+        console.log('Hello, World!');
+    }
+}
+const sayHello = returnFunc();
+sayHello();
+
+/**
+ * 4. **Callback Functions**
+    - Write a function that takes a callback and calls it with a sample argument.
+    - Explanation: Teaches about callback functions and how they work.
+ */
+function sum(a, b) {
+    return a + b;
+}
+function calculate(a, b, sum) {
+    return sum(a, b);
+}
+console.log(calculate(1, 2, sum));
+
+/**
+ * 5. **Array.filter Usage**
+    - Use `filter` to create a new array with elements that are numbers from the original array.
+    - Example input: `[1, 'a', 2]`
+    - Example output: `[1, 2]`
+    - Explanation: Introduces the `filter` method.
+ */
+function filterNumber(arr) {
+    return arr.filter((value) => typeof value === 'number' && !isNaN(value));
+}
+console.log(filterNumber([1, 'a', 2]));
+
+/**
+ * 6. **Array.reduce Method**
+    - Use `reduce` to sum all the numbers in an array.
+    - Example input: `[1, 2, 3, 4]`
+    - Example output: `10`
+    - Explanation: Teaches the `reduce` method for arrays.
+ */
+function returnTotal(arr) {
+    return arr.reduce((total, currentValue) => total + currentValue);
+}
+console.log(returnTotal([1, 2, 3, 4]));
+
+/**
+ * 7. **Creating and Using a Class**
+    - Create a class `Car` with properties like `model` and `year`, and a method that logs the model to the console.
+    - Explanation: Introduces classes and methods.
+
+ */
+class Car {
+    constructor(model, year) {
+        this.model = model;
```

Comment:
Em có thể mở rộng thêm chỗ này về keyword this trong JavaScript https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this